// Describes an authentication token
message AuthenticationToken {
  // The token name, a combination of the auth device and unique key.
  required string id = 1;

  // The user owning the token.
  optional string username = 2;

  // The date the token was created.
  required fixed32 created_time = 3;

  // The date after which the token is invalid, if any.
  optional fixed32 expire_time = 4;

  // Whether the token is enabled.
  required bool enabled = 5 [default=true];
}

// Describes a named beer variety, such as "India Pale Ale".
message BeerStyle {

  // The unique identifier for this beer style.
  required string id = 1;

  // The name of the beer style.
  required string name = 2;

}

// Describes a beer's brewer, style, and contents.
message BeerType {

  // The unique identifier for this beer type.
  required string id = 1;

  // The brand name of the beer.
  required string name = 2;

  // Brewer information for the beer.  May refer to an 'unknown' or generic
  // brewer record.
  required string brewer_id = 3;

  // Style information for the beer.  May refer to an 'unknown' or generic beer
  // style.
  required string style_id = 5;

  // For seasonal or special edition beers, the year or other edition name.
  optional string edition = 12;

  // Number of calories per ounce of beverage.
  optional float calories_oz = 7;

  // Number of carbohydrates per ounce of beverage.
  optional float carbs_oz = 8;

  // Alcohol by volume, as a percentage of the total volume.
  optional float abv = 9;

  // Original gravity of the beer, if known.
  optional float original_gravity = 10;

  // Specific/final gravity of the beer, if known.
  optional float specific_gravity = 11;

}

// Describes a specific producer of our favorite beverage.
message Brewer {

  // The unique identifier for this beer type.
  required string id = 1;

  // The name of the brewer.
  required string name = 2;

  // The country of the brewer's headquarters.
  optional string country = 3 [default = ''];

  // The state of the brewer's headquarters.
  optional string origin_state = 4;

  // The city of the brewer's headquarters.
  optional string origin_city = 5;

  // Type of production (usually either 'retail' or 'homebrew').
  optional string production = 6;

  // URL of brewer.
  optional string url = 7;

  // Free-form description.
  optional string description = 8;

}

// Describes a single recorded pour from the Kegbot.
message Drink {

  // The unique identifier for this drink.
  required uint32 id = 1;

  // The number of meter ticks recorded for the drink.
  required uint32 ticks = 2;

  // The volume of the drink, in milliliters.
  required float volume_ml = 3;

  // The session this drink belongs to.
  required string session_id = 4;

  // Local time when the drink was poured, in seconds since epoch (unix timestamp).
  required fixed32 pour_time = 5;

  // True if the drink is considered 'valid'.
  required bool is_valid = 6 [default = true];

  // The Keg from which the drink was poured.  May be unset if the drink could
  // not be associated with a keg.
  optional uint32 keg_id = 7;
  optional Keg keg = 8;

  // The User that poured the drink.  May be unset if the drinker was unknown.
  optional string user_id = 9;
  optional User user = 10;

}

// Describes a single instance of a Keg that is or was attached to the Kegbot.
message Keg {

  // The unique identifier for this keg.
  required uint32 id = 1;

  // The keg's current status.  Typically either "online" or "offline".
  required string status = 2;

  // The kind of beer within the keg.
  required string type_id = 3;

  // The size of the keg.
  required uint32 size_id = 4;

  // Local time when the keg was 'started', in seconds since epoch (unix
  // timestamp). This should be no later than the time of the first Drink.
  required fixed32 started_time = 5;

  // Local time when the keg was 'completed', in seconds since epoch (unix
  // timestamp). This should be no sooner than the time of the most recent
  // drink.  If the keg's `status` is not 'offline', then this value is
  // meaningless.
  required fixed32 finished_time = 6;

  // The adminstrator's description of this keg.
  optional string description = 7;

  // The percentage of the original volume that is remaining.
  required float percent_full = 8;

  // The volume, in milliliters, remaning.  This is simply a convenience value
  // that matches pct_full * size.volume_ml.
  required float volume_ml_remain = 9;

}

message KegSize {

  // The unique identifier for this keg size.
  required uint32 id = 1;

  // The name of the size ("Half Barrel", "Pony keg", ...)
  required string name = 2;

  // The volume of the size, in milliliters.
  required float volume_ml = 3;

}

// Describes a tap which is available for pouring beer.
message KegTap {

  // The unique identifier for this tap.
  required string id = 1;

  // The name of the tap, a free-form string ("Main tap", "Wet bar", ...)
  required string name = 2;

  // The meter the tap is assigned to.
  required string meter_name = 3;

  // Size of measurement.
  required float ml_per_tick = 4;

  // A longer description of the tap.
  optional string description = 5;

  // The Keg currently assigned to the tap.  May be unset if there is no keg
  // configured.
  optional uint32 current_keg_id = 6;

  // The temperature sensor for the tap, if any.
  optional string thermo_sensor_id = 7;

}

message Session {
  required string id = 1;
  required fixed32 start_time = 2;
  required fixed32 end_time = 3;
  required float volume_ml = 4;
}

message ThermoLog {
  required string id = 1;

  required string sensor_id = 2;

  // The temperature, in degrees C.
  required float temperature_c = 3;

  // The recording date.
  required fixed32 record_time = 4;
}

message ThermoSensor {
  // The raw/unique name of the sensor.
  required string sensor_name = 1;

  // The friendly name for the sensor.
  required string nice_name = 2;
}

// A drinker registered in the kegbot system.
message User {

  // The user's unique username.
  required string username = 1;

  // True if the user is active.  This value will be false for accounts which
  // have been disabled by.
  required bool is_active = 2 [default = true];

  // A URL to the profile picture of the user.  May be unset if the user does
  // not have a profile picture.
  optional string mugshot_url = 3;

  // True if the user is a member of the system's staff.
  required bool is_staff = 4 [default = false];

  // True if the user is an administrator.
  required bool is_superuser = 5 [default = false];

  // Local time for the user's registration, in seconds since epoch (unix
  // timestamp).
  required fixed32 joined_time = 6;

}

message UserSession {
  required string id = 1;
  required string session_id = 2;
  required string username = 3;
  required string start_time = 4;
  required string end_time = 5;
  required float volume_ml = 6;
}

message UserStats {
}
